### ISSUE TITLE
Multiple channels from the same client unreliable

### ISSUE THREAD

**Issue by haxsaw** [paramiko/paramiko#709]
It seems that if one creates multiple Channels from the same SSHClient instance and uses each channel independently in a separate thread, the results are unpredictable.

I developed the following example that illustrates what I'm seeing in my larger system:

```
import threading
import time
from paramiko import (SSHClient, AutoAddPolicy)


def do_command(sh, cmd="ls -l"):
    time.sleep(0.1)
    sh.send("%s\n" % cmd)
    time.sleep(0.1)
    result = sh.recv(4096)
    print(""">>>>>>>>>>>>>>>Start results
%s
<<<<<<<<<<<<End results""" % result)
    
    
if __name__ == "__main__":
    client = SSHClient()
    client.set_missing_host_key_policy(AutoAddPolicy())
    client.connect(hostname="localhost", username="lxle1",
                   key_filename="./lxle1-dev-key")
    threads = []
    for i in [1, 2]:
        sh = client.invoke_shell()
        t = threading.Thread(target=do_command, args=(sh,))
        threads.append(t)
        t.start()
        
    for t in threads:
        t.join()
```
The results of running this are unpredictable and sometimes surprising. A successful run can look like this:

```
(act)tom@maxwell:~/git/actuator/src/experiments$ python multishell.py 
>>>>>>>>>>>>>>>Start results
Welcome to Ubuntu 14.04.4 LTS (GNU/Linux 3.13.0-83-generic x86_64)

 * Documentation:  https://help.ubuntu.com/

24 packages can be updated.
0 updates are security updates.

Last login: Mon Mar 21 11:26:16 2016 from localhost
lxle1@maxwell:~$ ls -l
total 36
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:05 Desktop
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:58 Documents
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:05 Downloads
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:05 Music
drwxr-xr-x 3 lxle1 lxle1 4096 Mar  4 09:05 Pictures
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:05 Public
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:05 Templates
drwxrwxr-x 3 lxle1 lxle1 4096 Mar 17 08:45 tmp
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:05 Videos
lxle1@maxwell:~$ 
<<<<<<<<<<<<End results
>>>>>>>>>>>>>>>Start results
Last login: Mon Mar 21 11:26:19 2016 from localhost
lxle1@maxwell:~$ ls -l
total 36
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:05 Desktop
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:58 Documents
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:05 Downloads
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:05 Music
drwxr-xr-x 3 lxle1 lxle1 4096 Mar  4 09:05 Pictures
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:05 Public
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:05 Templates
drwxrwxr-x 3 lxle1 lxle1 4096 Mar 17 08:45 tmp
drwxr-xr-x 2 lxle1 lxle1 4096 Mar  4 09:05 Videos
lxle1@maxwell:~$ 
<<<<<<<<<<<<End results
```

Note the missing MOTD/herald from the second output.

An unsuccessful run can result in a few different tracebacks, but the following in typical:

```
(act)tom@maxwell:~/git/actuator/src/experiments$ python multishell.py 
Traceback (most recent call last):
  File "multishell.py", line 23, in <module>
    sh = client.invoke_shell()
  File "/home/tom/vpy/act/local/lib/python2.7/site-packages/paramiko/client.py", line 429, in invoke_shell
    chan.get_pty(term, width, height, width_pixels, height_pixels)
  File "/home/tom/vpy/act/local/lib/python2.7/site-packages/paramiko/channel.py", line 60, in _check
    return func(self, *args, **kwds)
  File "/home/tom/vpy/act/local/lib/python2.7/site-packages/paramiko/channel.py", line 177, in get_pty
    self._wait_for_event()
  File "/home/tom/vpy/act/local/lib/python2.7/site-packages/paramiko/channel.py", line 1086, in _wait_for_event
    raise e
EOFError
Exception in thread Thread-3:
Traceback (most recent call last):
  File "/usr/lib/python2.7/threading.py", line 810, in __bootstrap_inner
    self.run()
  File "/usr/lib/python2.7/threading.py", line 763, in run
    self.__target(*self.__args, **self.__kwargs)
  File "multishell.py", line 8, in do_command
    sh.send("%s\n" % cmd)
  File "/home/tom/vpy/act/local/lib/python2.7/site-packages/paramiko/channel.py", line 698, in send
    return self._send(s, m)
  File "/home/tom/vpy/act/local/lib/python2.7/site-packages/paramiko/channel.py", line 1058, in _send
    raise socket.error('Socket is closed')
error: Socket is closed
```

Oddly, it is generally the first run that fails in a group of runs, and then success and failure appear at random times when running the prog repeatedly.

I have yet to observe this behavior if each channel is created on a different SSHClient instance. Since channel creation seems to be much faster than SSHClient creation, I was hoping to be able to use the same client and just create all the channels I needed for parallel operations. It appears the work around is to always make new connections at the performance cost, but at least things are stable.

Is this a Paramiko issue or something with the underlying SSH protocol?

**Comment by radssh** [comment]
The MOTD only appearing once is a quirk of Ubuntu using the pam_motd.so module (See /etc/pam.d/sshd) and disabling PrintMotd in /etc/ssh/sshd_config. It is PAM, not sshd directly that includes the motd content, and it only is sent on the initial authentication, not for all channels belonging to the same underlying transport. If you had content in a static /etc/motd file, and enabled PrintMotd in sshd_config, you would see its content appear for all channels (including the first channel getting it twice; once from pam, and again from sshd itself).

The other issue appears to be related to https://bugs.launchpad.net/ubuntu/+source/openssh/+bug/1334916 bug in sshd that looks like it may not (yet) be available under Ubuntu 14.04. When I was able to reproduce in a 14.04 VM, I was seeing this in the /var/log/auth.log file:
```
Apr  1 16:23:36 trusty sshd[26164]: fatal: mm_request_receive_expect: read: rtype 157 != type 29
Apr  1 16:23:36 trusty sshd[26173]: fatal: mm_request_receive_expect: read: rtype 29 != type 157
```
Where 157 is a ConsoleKit answer message ID, and 29 is a RequestPTY answer message, that seem to be getting mismatched to their requestors, in turn causing the sshd server side to drop your connection.

**Comment by haxsaw** [comment]
Thanks for this analysis. Sadly, it does't appear to be fixed in Ubuntu 15.10 either. I just tried my above program out on a machine I recently put together with 15.10 and while it **is** more stable, you still get the occasional error with the same messages in auth.log you note above. To ensure that the problem isn't Paramiko's, I built the example C program found at the Launchpad bug you noted above, and it still errors out. Running apt-show-versions shows that my up-to-date Ubuntu 15.10 has an earlier version of openssh than has the fix for this bug. I guess I'm stuck with my workaround until this gets pushed out. Thanks again.
